### PR TITLE
Spec freeze updates

### LINKED ISSUE #2307: [Mega Tracking] - Align spec to release v0.6
This PR tracks the overall progress of aligning current beacon chain implementation to spec version 0.6. Please comment on the issue before you pick up an item to work on. 

- [x] 💎  **Update [constant](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#constants) names and values** 💎
- [x] 💎  **Update Protobuf data structures** 💎
    - [x] Update [Crosslink](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#crosslink) | #2313
    - [x] Update [Eth1Data](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#eth1data) | #2315
    - [x] Update [AttestationData](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attestationdata) | #2316
    - [x] Update [AttestationDataAndCustodyBit](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attestationdataandcustodybit) | #2316
    - [x] Update [IndexedAttestation](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#indexedattestation) | #2316
    - [x] Update [DepositData](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#depositdata) | #2318
    - [x] Update [BeaconBlockHeader](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#beaconblockheader) | #2327
    - [x] Update [Validator](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#validator) | #2318
    - [x] Update [HistoricalBatch](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#historicalbatch) | #2327
    - [x] Update [ProposerSlashing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#proposerslashing) | #2328
    - [x] Update [AttesterSlashing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attesterslashing) | #2328
    - [x] Update [Attestation](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#Attestation) | #2328
    - [x] Update [Deposit](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#Deposit) | #2327
    - [x] Update [VoluntaryExit](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#VoluntaryExit) | #2328
    - [x] Update [Transfer](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#Transfer) | #2328
    - [x] Update [BeaconBlockBody](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#BeaconBlockBody) | #2340
    - [x] Update [BeaconBlock](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#BeaconBlock) | #2340
    - [x] Update [BeaconState](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#BeaconState) | #2341 
    - [x] Update [Fork](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#fork) | #2545

- [x] 💎  **Update general [helper functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#helperfunctions)**  💎
    - [x] Update slot related helpers | #2345 
    - [x] Update validator related helpers | #2352 
    - [x] Update balance related helpers | #2348 
    - [x] Update permuted index helpers | #2350
    - [x] Update splitoffset helpers | #2365
    - [x] Update committee related helpers | #2354
    - [x] Update shard delta helpers | #2497
    - [x] Update epoch start shard helpers | #2497
    - [x] Update block root related helpers | #2355
    - [x] Update state root related helpers | #2355
    - [x] Update randao mix helpers |  #2357
    - [x] Update active index root helpers | #2369
    - [x] Update generate seed helpers | #2371
    - [x] Update beacon proposer index helpers | #2381
    - [x] Update attesting indices helpers | #2449
    - [x] Update balance related helpers | #2356
    - [x] Update bitfield related helpers | #2446 (bug fix)
    - [x] Update verify indexed attestation helpers | #2385
    - [x] Update churn limit helpers | #2364
    - [x] Update [convert to indexed ](https://github.com/ethereum/eth2.0/specs/blob/dev/specs/core/0_beacon-chain.md#convert_to_indexed) | #2519 
    - [x] update [get domain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_domain) | #2545

- [x] 💎  **Update routines for [changing validator status](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#routines-for-updating-validator-status)****  💎
    - [x] Update initiate validator exit | #2366 
    - [x] Update slash validator (relies on balance and BeaconProposerIndex changes) | #2391

- [x] 💎  **Verify and update [chain start logic](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#eth2genesis-log)**  💎
    - [x] Update [genesis functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#on-genesis) | #2465 
    - [x] Update Deposit Contract to new spec.

- [x] 💎  **Verify and update [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md)** 💎 
    - [x] 🔹 Implement [self-signed containers](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md#self-signed-containers)🔹

- [x] 💎  **Update beacon chain [state transition function](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#beacon-chain-state-transition-function)**  💎
    - [x] 🔹 Update [per epoch processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#per-epoch-processing)🔹
        - [x] Update [general helper functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#helper-functions-1) | #2379
        - [x] Update [justification and finalization functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#justification-and-finalization) | #2448
        - [x] Update [crosslink function](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#crosslinks) | #2460 
        - [x] Update [rewards and penalties functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#rewards-and-penalties) | #2517
        - [x] Update [registry updates functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#registry-updates) | #2521
        - [x] Update [slashings function](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#slashings)
        - [x] Update [final updates functions](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#final-updates)
    - [x]  🔹 Update [per slot processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#per-slot-processing)🔹
    - [x]  🔹 Update [per block processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#per-block-processing)🔹
        - [x] Implement [process block header](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#block-header) | #2534
        - [x] Update [process randao](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#randao) | #2614
        - [x] Update [process eth1 data](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#eth1-data)
        - [x] Update [proposer slashing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#proposer-slashings) | #2603
        - [x] Update [attester slashing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attester-slashings)
        - [x] Update [attestation processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#attestations)
        - [x] Update [deposits processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#deposits)
        - [x] Update [exits processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#voluntary-exits)
        - [x] Implement [transfers processing](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#transfers)
        - [x] Update [state root verification](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#state-root-verification)

- [x] 💎  **Verify and update [BLS](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/bls_signature.md#bls-signature-verification)
     - [x] 🔹  add [bls verify](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/bls_signature.md#bls_verify)

- [x] align RPC endpoint [eth1data to spec](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/validator/0_beacon-chain-validator.md#eth1-data) | #2733

### PR DESCRIPTION
**Pull request by prestonvanloon**
Resolves #2307

Merge all v0.6+ features into this branch. This can be reviewed and merged when all boxes of #2307 are checked. 

List your blockers here:
- [x] Remove spec test yaml from repo, use starlark & bazel to fetch from remote repository. Owner=preston
- [x] All spec tests passing #2812 
- [x] SSZ compliance https://github.com/prysmaticlabs/go-ssz/pull/41

### CODE REVIEW

File: shared/trieutil/sparse_merkle.go [hunk 1]
```diff
@@ -73,7 +100,22 @@ func BranchIndices(merkleIndex int, depth int) []int {
 	return indices
 }
 
-// Root of the Merkle trie.
+// HashTreeRoot of the Merkle trie as defined in the deposit contract.
+//  Spec Definition:
+//   sha256(concat(node, self.to_little_endian_64(self.deposit_count), slice(zero_bytes32, start=0, len=24)))
```
Comment (rauljordan) [inline shared/trieutil/sparse_merkle.go]: i don't believe we need this - we should simply compute hash tree root with capacity of the list of deposit data instead

File: tools/ssz-server/main.go [hunk 2]
```diff
@@ -5,7 +5,6 @@
 package main
```
Comment (rauljordan) [inline tools/ssz-server/main.go]: We should open an issue to remove this

File: beacon-chain/core/blocks/block.go [hunk 3]
```diff
@@ -17,64 +17,49 @@ var clock utils.Clock = &utils.RealClock{}
...
-		if earliestSlot < params.BeaconConfig().GenesisSlot {
-			earliestSlot = params.BeaconConfig().GenesisSlot
-		}
-		return []byte{}, fmt.Errorf("slot %d is not within expected range of %d to %d",
-			slot-params.BeaconConfig().GenesisSlot,
-			earliestSlot-params.BeaconConfig().GenesisSlot,
-			state.Slot-params.BeaconConfig().GenesisSlot,
-		)
+// BlockFromHeader manufactures a block from its header. It contains all its fields,
+// expect for the block body.
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block.go]:
```suggestion
// except for the block body.
```

File: shared/p2p/options.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ import (
 	peer "github.com/libp2p/go-libp2p-peer"
 	filter "github.com/libp2p/go-maddr-filter"
 	ma "github.com/multiformats/go-multiaddr"
+
```
Comment (prestonvanloon) [inline shared/p2p/options.go]:
```suggestion
```

File: beacon-chain/utils/spectest/shuffle_test_format.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package spectest
```
Comment (prestonvanloon) [inline beacon-chain/utils/spectest/shuffle_test_format.go]: This shouldn't be a public target. Lets rename to shuttle_test_format_test.go to keep private to this.

**Comment by prestonvanloon** [comment]
Blocked by #2959

**Comment by prestonvanloon** [comment]
At runtime, all genesis validators are reporting status of PENDING_ACTIV.

File: beacon-chain/core/state/transition.go [hunk 6]
```diff
@@ -77,103 +79,134 @@ func ExecuteStateTransition(
...
+//  def process_slots(state: BeaconState, slot: Slot) -> None:
+//    assert state.slot <= slot
+//    while state.slot < slot:
+//        process_slot(state)
+//        # Process epoch on the first slot of the next epoch
+//        if (state.slot + 1) % SLOTS_PER_EPOCH == 0:
+//            process_epoch(state)
+//        state.slot += 1
+//    ]
+func ProcessSlots(ctx context.Context, state *pb.BeaconState, slot uint64) (*pb.BeaconState, error) {
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/core/state/transition.go]: What is the suggestion? We need a simple API to skip slots

File: tools/ssz-server/main.go (hunk 2 above)
Comment (prestonvanloon) [reply tools/ssz-server/main.go]: We use it for the testnet site.

File: beacon-chain/blockchain/block_processing_test.go [hunk 7]
```diff
@@ -27,8 +29,15 @@ import (
 // Ensure ChainService implements interfaces.
 var _ = BlockProcessor(&ChainService{})
 
+func init() {
+	// TODO(2993): remove this after ssz is optimized for mainnet.
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/block_processing_test.go]: either format is OK

**Pull request merged**